Added HighMaps

Prakriti-nith opened the pull request:
Examples link: [1](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/0314f9e743a356b0551445e0bc8ae315718acd73/spec/dummy_iruby/HighMaps%20-%20Dynamic.ipynb), [2](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/0314f9e743a356b0551445e0bc8ae315718acd73/spec/dummy_iruby/HighMaps%20|%20General.ipynb), [3](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/0314f9e743a356b0551445e0bc8ae315718acd73/spec/dummy_iruby/HighMaps%20|%20Series%20Types.ipynb),

I am working on loading less js files.

Prakriti-nith commented:
Revised Examples (containing `modules` option): [link1](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/02237bf42dfed101b4d91de7fd08bc172cce2304/spec/dummy_iruby/HighMaps%20|%20Series%20Types.ipynb), [link2](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/02237bf42dfed101b4d91de7fd08bc172cce2304/spec/dummy_iruby/HighMaps%20|%20General.ipynb), [link3](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/02237bf42dfed101b4d91de7fd08bc172cce2304/spec/dummy_iruby/HighMaps%20-%20Dynamic.ipynb).
I have kept all the modules of HighCharts, HighStock and HighMaps in the /modules folder so that user only needs to specify the name of the module (like europe.js) and that will be loaded. I am loading the modules in `show_in_iruby` method before getting the script of the chart and it is working fine now.
For the web frameworks, `div` method is called. So I tried loading the dependent js in `to_html` method which is working correctly. My concern is that the dependent files of the modules are not loaded in the head tag instead they are loaded into the body before the script of the chart, so will it be a problem?

Prakriti-nith commented:
When the plotting_library (or dependent_script) is set to highcharts, the dependencies (highstock.js, map.js) and the dependencies of the basic modules that are common to all the three (HighCharts, HighStock, HighMaps) will be loaded. Then the user can provide the required modules of the chart in options. These will be loaded in `show_in_iruby` for iruby notebook and in `div` (i.e., in to_html) for web frameworks.

Shekharrajak commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -3,7 +3,7 @@
 module LazyHighCharts
   def self.init_script(
-    dependent_js=['highstock.js', 'highcharts-more.js', 'modules/exporting.js',
+    dependent_js=['highstock.js', 'map.js', 'modules/exporting.js',
```
Since it is constant we must place all the constant variable into one file, to make it loosely coupled. Accessing these constants into multiple files will be better.

Shekharrajak commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -61,6 +64,17 @@ def to_html_iruby(placeholder=random_canvas_id)
       high_chart_iruby(extract_chart_class, placeholder, self)
     end
 
+    # @return load the dependent modules of the chart
```
At least Input parameter and return object must be documented according to yard.

Prakriti-nith commented:
I have written the examples here: [link1](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/c2936058d05a18dbf17cbed0aa94ce3ee1810256/spec/dummy_iruby/HighMaps%20|%20General.ipynb), [link2](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/c2936058d05a18dbf17cbed0aa94ce3ee1810256/spec/dummy_iruby/HighMaps%20|%20Series%20Types.ipynb), [link3](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/c2936058d05a18dbf17cbed0aa94ce3ee1810256/spec/dummy_iruby/HighMaps%20-%20Dynamic.ipynb).

Prakriti-nith commented:
In this commit I have:
1. Loaded the mapdata for each chart directly from options[:chart][:map]. This option is **always** provided by the user which requires additional mapdata. It is actually the path of the dependent js file. So no need to explicitly provide it in `modules` option.
2. User can still provide `modules` option for highcharts modules (treemap.js, tilemap.js, sankey.js, etc) as modules/treemap, etc. I will add these dependencies and their examples in a separate PR.
3. Refactored rake task accordingly.
4. Added more examples of countries with data as Daru::DataFrame.
5. Kept the constants in a separate file.

Examples: [Countries](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/dc876d00b038d8f9eac4f65c1e7e70cfb43d9ebd/spec/dummy_iruby/HighMaps%20-%20Countries%20with%20data%20as%20DataFrame.ipynb), [General](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/dc876d00b038d8f9eac4f65c1e7e70cfb43d9ebd/spec/dummy_iruby/HighMaps%20|%20General.ipynb), [Series](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/dc876d00b038d8f9eac4f65c1e7e70cfb43d9ebd/spec/dummy_iruby/HighMaps%20|%20Series%20Types.ipynb), [Dynamic](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/dc876d00b038d8f9eac4f65c1e7e70cfb43d9ebd/spec/dummy_iruby/HighMaps%20-%20Dynamic.ipynb).
Please review this PR and let me know any changes that I have to make.

Prakriti-nith commented:
I tried using `data = series_dt` but it was not working as most of the advance highmaps use multiple series' or data in a specified format. I have included the examples of Daru::DataFrame [here](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/dc876d00b038d8f9eac4f65c1e7e70cfb43d9ebd/spec/dummy_iruby/HighMaps%20-%20Countries%20with%20data%20as%20DataFrame.ipynb).

Shekharrajak commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -1,10 +1,10 @@
   def self.init_script(
-    dependent_js=['highstock.js', 'highcharts-more.js', 'modules/exporting.js',
-                  'highcharts-3d.js', 'modules/data.js']
+    dependent_js=[HIGHSTOCK, MAP, EXPORTING, HIGHCHARTS_3D, DATA]
```
`[HIGHSTOCK, MAP, EXPORTING, HIGHCHARTS_3D, DATA]` is also constant as web app highcharts dependencies.

Prakriti-nith commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -61,6 +64,47 @@ def to_html_iruby(placeholder=random_canvas_id)
+      end
+      # Dependencies provided in modules option (of highcharts mainly
+      #   like tilemap) by the user
+      dep_js |= options.delete(:modules).map! { |js| "#{js}.js" } unless
```
`map` option is not created by us and will be provided in highcharts' options everytime there is a need for map data(like europe.js). This would be the exact address of the dependency required.
For example, a user using dependency `https://code.highcharts.com/mapdata/countries/in/in-all.js` will provide `countries/in/in-all` as a value for options[:chart][:map]. 
Only the options `chart_class` and `modules` are created by us.
Yes, we can have a third parameter for these but IMO that would be only more complex for the user plus we then have to generalize it for other adapters as well which I think is not a good option.

Prakriti-nith replied:
`modules` option is created just to load some additional dependencies used in a specific chart by the user.

Prakriti-nith commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -61,6 +64,47 @@ def to_html_iruby(placeholder=random_canvas_id)
+      dep_js = []
+      # Mapdata dependencies
+      if !options[:chart_class].nil? && options[:chart_class].capitalize == 'Map'
+        dep_js.push('mapdata/' + options[:chart][:map].to_s + '.js') if
```
I have kept the same folder structure provided in the highcharts dependencies. It is because the value provided by the user in options[:chart][:map] will be the same as of the dependencies and by this folder structure it will be easy to retrieve the data.

Prakriti-nith commented:
I have added [examples](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/c242a7cb856802a18b7639f486abe5d04674a392/spec/dummy_iruby/Highcharts%20-%20More%20Chart%20Types%20(Using%20modules).ipynb) of highcharts that uses modules option.

Prakriti-nith commented on lib/tasks/high_charts.rake:
```diff
@@ -57,4 +58,31 @@ namespace :highcharts do
+      sh "mkdir -p lib/daru/view/adapters/js/highcharts_js/mapdata/countries/"
+      sh "mkdir -p lib/daru/view/adapters/js/highcharts_js/mapdata/custom/"
+
+      countries_folders = ['in', 'us', 'gb', 'af', 'au', 'bd', 'be', 'ca', 'cn', 'fr']
```
I think around 10 - 15 MBs. Ok, I will download them all.

Prakriti-nith commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -61,6 +64,47 @@ def to_html_iruby(placeholder=random_canvas_id)
+      options[:modules].nil?
+      dep_js
+    end
+    # rubocop:enable Metrics/AbcSize
```
Yes, I tried removing it but the conditions are necessary. Please provide any suggestions to fix it.

Prakriti-nith replied:
I tried it earlier but still, there was rubocop error. I have separated out the method.

Prakriti-nith commented:
I have added the support for multiple series in highcharts. I have also updated all the examples of highcharts, highstock and highmaps using `Daru::View::Plot.new(data, opts)` and almost all the highstock examples using Daru::DataFrame, Daru::Vector or Array as `data`.

Prakriti-nith commented:
@Shekharrajak please review the changes.

Prakriti-nith commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -33,22 +33,25 @@ class HighChart
     #
     def to_html(placeholder=random_canvas_id)
       chart_hash_must_be_present
+      script = load_dependencies('web_frameworks')
```
I will look into the Proxy design. However, `to_html` will load the dependencies only if there is some additional dependency is provided in the `modules` option or if a map is being drawn (to load the required map data). Otherwise, `load_dependencies` will simply return an empty string.

Prakriti-nith replied:
`to_html` will load the dependencies only if there is some additional dependency provided in the modules option or if a map is being drawn (to load the required map data). Otherwise, load_dependencies will simply return an empty string. I will look into the Proxy design pattern also.

Shekharrajak replied:
Need to handle this.

Prakriti-nith commented on lib/daru/view/adapters/highcharts.rb:
```diff
@@ -42,10 +40,15 @@ def init(data=[], options={})
-            series_name = options[:name] unless options[:name].nil?
-            f.series(type: series_type, name: series_name, data: data_new)
+            # For multiple series
+            if data.is_a?(Array) && data[0].is_a?(Hash)
```
Multiple series are used in almost all the HighCharts and HighMaps examples like: ((link1)[http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/b219f80305530351e4de079ded8f109595012483/spec/dummy_iruby/HighCharts%20-%20Area%20chart.ipynb], (link2)[http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/b219f80305530351e4de079ded8f109595012483/spec/dummy_iruby/HighCharts%20-%20Bar%20Chart.ipynb], (link3)[http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/b219f80305530351e4de079ded8f109595012483/spec/dummy_iruby/HighMaps%20|%20Series%20Types.ipynb]). `guess_data` does not work for Array of Hashes (Multiple series) and will raise ArgumentError. Our data is already in the series format and can be used directly.

Prakriti-nith replied:
Multiple series are used in almost all the HighCharts and HighMaps examples like: ([link1](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/b219f80305530351e4de079ded8f109595012483/spec/dummy_iruby/HighCharts%20-%20Area%20chart.ipynb), [link2](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/b219f80305530351e4de079ded8f109595012483/spec/dummy_iruby/HighCharts%20-%20Bar%20Chart.ipynb), [link3](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/b219f80305530351e4de079ded8f109595012483/spec/dummy_iruby/HighMaps%20|%20Series%20Types.ipynb)). `guess_data` does not work for Array of Hashes (Multiple series) and will raise ArgumentError. Our data is already in the series format and can be used directly.

Shekharrajak commented on lib/daru/view/adapters/highcharts.rb:
```diff
@@ -42,10 +40,17 @@ def init(data=[], options={})
+            # For multiple series when data is in a series format as in
+            # HighCharts official examples
+            # TODO: Add support for multiple series when data as Daru::DataFrame/Daru::Vector
+            if data.is_a?(Array) && data[0].is_a?(Hash)
```
It is already working for multiple series, since it is checking for array in `guess_data ` , right ?

Shekharrajak commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -61,6 +64,54 @@ def to_html_iruby(placeholder=random_canvas_id)
+      options[:chart] &&
+      options[:chart][:map]
+    end
+    # rubocop:enable Metrics/AbcSize
```
remove this line.

Prakriti-nith replied:
IMO all the four conditions are necessary. Which one should I remove?

Shekharrajak commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -61,6 +64,54 @@ def to_html_iruby(placeholder=random_canvas_id)
+    # @return [void] Appends the map data in dep_js
+    # rubocop:disable Metrics/AbcSize
+    def get_map_data_dependencies(dep_js)
+      dep_js.push('mapdata/' + options[:chart][:map].to_s + '.js') if
```
````
if (condition)
  .. statements... 
  ....

````
is easy to understand rather than this.

Shekharrajak commented:
I want to discuss about the future implementation (that we can do to improve it),if anything in TODO for map part. 

Update them here:  https://github.com/SciRuby/daru-view/wiki/HighCharts-features

Prakriti-nith commented:
I wanted to edit the wiki online but it says `You do not have permission to update this wiki`. Please provide the required permissions.

Shekharrajak commented:
Wait, I didn't know about that. It will be public soon.

Prakriti-nith commented:
@Shekharrajak  I have made the required changes. Please review.

Shekharrajak commented:
Please add some solid examples that can show the full features of daru-view/highcharts/highmap in new wiki page ( Starting from easy to advance). It will be very useful , since there are lot of new things added in this PR. 

Once you resolve the conflicts, it will be merged.
Thanks!

Shekharrajak commented on lib/daru/view/adapters/highcharts/display.rb:
```diff
@@ -61,6 +64,52 @@ def to_html_iruby(placeholder=random_canvas_id)
+
+    # @param dep_js [Array] JS dependencies required for drawing a map(mapdata)
+    # @return [void] Appends the map data in dep_js
+    def get_map_data_dependencies(dep_js)
```
According to HighMap option `map`, we are loading the respective js files (so no need to load not used files). 

But need to find a way to capture the loaded js files, so that next time we run `to_html` or generate the script it won't load the file, that is already loaded into the **same  web-page**.

Shekharrajak commented on lib/tasks/high_charts.rake:
```diff
@@ -57,4 +63,55 @@ namespace :highcharts do
     end
   end
 
+  task :map do
```
Need to find a way to handle many js files . It is increasing the gem size.

Shekharrajak commented:
Noting down few important points, for future plan : 

1. [load dependencies ](https://github.com/SciRuby/daru-view/pull/92/files#r192598853)
2. [loading map js file](https://github.com/SciRuby/daru-view/pull/92#discussion_r195900248)
2. [Need to handle many js files](https://github.com/SciRuby/daru-view/pull/92#discussion_r195900260)

Anything you want to add here , @Prakriti-nith ?
